packages/cloud-functions/src/cico/Moonpay.ts

```diff
@@ -0,0 +1,144 @@
+    supportsAddressTag: boolean
+    addressTagRegex: string | null
+    supportsTestMode: boolean
+    isSuspended: boolean
+    isSupportedInUS: boolean
+    isSellSupported: boolean
+    notAllowedUSStates: string[]
+  }
+  extraFeeAmount: number
+  feeAmount: number
+  networkFeeAmount: number
+  paymentMethod: string
+  quoteCurrencyAmount: number
+  totalAmount: number
+}
+
+const Moonpay = {
+  fetchQuote: async (
+    digitalAsset: DigitalAsset,
+    fiatCurrency: string,
+    fiatAmount: number | undefined,
+    userLocation: UserLocationData
+  ) => {
+    try {
+      if (!fiatAmount) {
+        throw Error('Purchase amount not provided')
+      }
+
+      let { localCurrency, exchangeRate } = await fetchLocalCurrencyAndExchangeRate(
+        userLocation.country,
+        fiatCurrency
+      )
+
+      // If the local currency is not supported by Moonpay, then get estimate in USD
+      if (!MOONPAY_DATA.supported_currencies.includes(localCurrency)) {
+        ;({ localCurrency, exchangeRate } = await fetchLocalCurrencyAndExchangeRate(
+          userLocation.country,
+          fiatCurrency,
+          'USD'
+        ))
+      }
+
+      const localFiatAmount = fiatAmount * exchangeRate
+
+      const paymentMethods = ['credit_debit_card']
+      if (userLocation.country) {
+        if (bankingSystemToCountry.gbp[userLocation.country]) {
+          paymentMethods.push('gbp_bank_transfer')
+        } else if (bankingSystemToCountry.sepa[userLocation.country]) {
+          paymentMethods.push('sepa_bank_transfer')
+        }
+      }
+
+      const baseUrl = `
+        ${MOONPAY_DATA.api_url}
+        /v3
+        /currencies
+        /${digitalAsset.toLowerCase()}
+        /buy_quote
+        /?apiKey=${MOONPAY_DATA.public_key}
+        &baseCurrencyCode=${localCurrency.toLowerCase()}
+        &baseCurrencyAmount=${localFiatAmount.toFixed(2)}
+      `.replace(/\s+/g, '')
+
+      const rawQuotes: Array<MoonpayQuote | null> = await Promise.all(
+        paymentMethods.map((method) => Moonpay.get(`${baseUrl}&paymentMethod=${method}`))
+      )
+
+      const quotes: ProviderQuote[] = []
+      for (const quote of rawQuotes) {
+        if (!quote) {
+          continue
+        }
+
+        const paymentMethod =
+          quote.paymentMethod === 'credit_debit_card' ? PaymentMethod.Card : PaymentMethod.Bank
+
+        const { feeAmount, extraFeeAmount, networkFeeAmount } = quote
+        quotes.push({
+          paymentMethod,
+          fiatFee: (feeAmount + extraFeeAmount + networkFeeAmount) / exchangeRate,
+          returnedAmount: quote.quoteCurrencyAmount,
+          digitalAsset: quote.currency.code,
+        })
+      }
+
+      if (!quotes.length) {
+        return
+      }
+
+      return quotes
+    } catch (error) {
+      console.error('Error fetching Moonpay quote: ', error)
+    }
+  },
+  get: async (path: string) => {
+    try {
+      const response = await fetchWithTimeout(path, null, FETCH_TIMEOUT_DURATION)
```

Comment:
All the callers of `fetchWithTimeout` look like they have a similar pattern: call, check if response is falsey, json parse.  Is it worth pulling that pattern into `fetchWithTimeout` or something else?

Reply:
The original idea was to make `fetchWithTimeout` a 1-to-1 replacement for `fetch` so it would feel familiar to future developers that want to use it. I can go either way on the potential usability benefits outweighing the DRY benefits

Reply:
I can at least simplify it a little bit by bundling the falsy response check into `fetchWithTimeout`